### PR TITLE
Feature enc x24 j600 wip

### PR DESCRIPTION
**Pull request by asifsardar26**
- Adding classes Register_GPIO_BSRRL & Register_GPIO_BSRRH and modifying GPIO.h
- Adding objects for Register_GPIO_BSRRL & Register_GPIO_BSRRH in emulator.mk
- Now, all the RF tests works fine with ellduino, stm32f4discovery & stm32f334nucleo

### CODE REVIEW

File: stm32/cores/arduelli/Arduino_Serial.h [hunk 1]
```diff
@@ -59,7 +59,7 @@ class SerialClass : public Stream {
 constexpr SerialClass::SerialClass(const USART &usart)
     : usart_(usart) {}
 
-ARDUINO_INLINE_MEMBER_FUNCTION
+inline
```
Comment (pekkanikander) [inline stm32/cores/arduelli/Arduino_Serial.h]: Please don't do these changes of replacing the macros with plain keywords.  The macros are useful as documentation, even though they always expand in the same way.

File: stm32/emulator/inc/GPIO.h [hunk 2]
```diff
@@ -24,11 +24,19 @@
 #include <Register_GPIO_PUPDR.h>
 #include <Register_GPIO_IDR.h>
 #include <Register_GPIO_ODR.h>
-#include <Register_GPIO_BSRR.h>
-//#include <Register_GPIO_LCKR.h>
-#include <Register_GPIO_AFR.h>
-#include <Register_GPIO_BRR.h>
-
+#if defined(__STM32F407__) || defined(__STM32F334__)
+    #include <Register_GPIO_BSRRL.h>
```
Comment (pekkanikander) [inline stm32/emulator/inc/GPIO.h]: When intending pre-processor directives, don't intend the whole line, but leave the # sign in the beginning of the line, and intend with just a single space per intending level.

File: stm32/emulator/inc/GPIO.h [hunk 3]
```diff
@@ -79,4 +92,4 @@ GeneralPurposeInputOutput *const GPIOE = &GeneralPurposeInputOutput::GPIOE;
 GeneralPurposeInputOutput *const GPIOF = &GeneralPurposeInputOutput::GPIOF;
 GeneralPurposeInputOutput *const GPIOG = &GeneralPurposeInputOutput::GPIOG;
 GeneralPurposeInputOutput *const GPIOH = &GeneralPurposeInputOutput::GPIOH;
-GeneralPurposeInputOutput *const GPIOI = &GeneralPurposeInputOutput::GPIOI;
+GeneralPurposeInputOutput *const GPIOI = &GeneralPurposeInputOutput::GPIOI;
```
Comment (pekkanikander) [inline stm32/emulator/inc/GPIO.h]: Avoid these kinds of spurious changes.  The last line in the file should end with a newline.

File: stm32/emulator/inc/Register_GPIO_ODR.h [hunk 4]
```diff
@@ -39,6 +41,8 @@ class Register_GPIO_ODR : public Register {
         {}
     uint32_t  operator = (uint32_t arg);
 private:
+	friend class Register_GPIO_BSRRL;
```
Comment (pekkanikander) [inline stm32/emulator/inc/Register_GPIO_ODR.h]: Fix intending here.

File: stm32/emulator/src/Register_GPIO_BSRRH.cpp [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+#include <Register_GPIO_ODR.h>
+
+uint32_t  Register_GPIO_BSRRH::operator = (uint32_t arg) {
+    printout("=", 0);
+
+    uint32_t value = odr_.value_;
+    value &= ~(arg & 0xffff);
+    odr_.assign(value);
+
+    return 0; // See STM32F0xx Refence Manual RM0091 Doc ID 018940 Rev 5 page 162.
```
Comment (pekkanikander) [inline stm32/emulator/src/Register_GPIO_BSRRH.cpp]: Please give references to the right docs, i.e. STM32F4 and STM32F3 manuals.

File: stm32/make/emulator.mk [hunk 6]
```diff
@@ -25,6 +25,7 @@ EXTRA_CFLAGS += \
 	-I$(TOP)system/stm32/inc \
 	-I$(TOP)system/stm32/CMSIS/Include \
 	-I$(TOP)variants/$(VARIANT) \
+	-I$(TOP)libraries/SPI/src/variants \
```
Comment (pekkanikander) [inline stm32/make/emulator.mk]: Why is this needed?  I'm not saying it is not, but I'm unsure.  Therefore I need to know the reason.

File: stm32/make/system_libs_inc.mk [hunk 7]
```diff
@@ -5,6 +5,18 @@
...
 #
+# Define the system libraries.  Add a new one for each variant.
+# If you update these here, please be sure to make the same
+# change into the VARIANTS in system_libs.mk
+#
+SYSTEM_LIBS := \
+    $(TOP)variants/ellduino/libstm32f0.a \
+    $(TOP)variants/stm32f334nucleo/libstm32f3.a \
+    $(TOP)variants/stm32f4discovery/libstm32f4.a
+#   $(TOP)variants/stm32f0discovery/libstm32f0.a \
```
Comment (pekkanikander) [inline stm32/make/system_libs_inc.mk]: Again, I'm not sure why exactly this is needed.  I think it is not.

File: stm32/make/system_libs_inc.mk [hunk 8]
```diff
@@ -5,6 +5,18 @@
...
+
+
+#
```
Comment (pekkanikander) [inline stm32/make/system_libs_inc.mk]: This is a spurious change.  Remove

File: stm32/system/stm32/src/startup_stm32f334x8.S [hunk 9]
```diff
@@ -0,0 +1,421 @@
+/**
```
Comment (pekkanikander) [inline stm32/system/stm32/src/startup_stm32f334x8.S]: Please remove this file.  It is no longer needed.

File: stm32/system/stm32/src/system_stm32f3xx.c [hunk 10]
```diff
@@ -200,6 +200,8 @@ void SystemInit(void)
 
 #ifdef VECT_TAB_SRAM
   SCB->VTOR = SRAM_BASE | VECT_TAB_OFFSET; /* Vector Table Relocation in Internal SRAM */
+#else
```
Comment (pekkanikander) [inline stm32/system/stm32/src/system_stm32f3xx.c]: This was removed earlier on by Ivan, I believe.  Please check.

File: stm32/tests/robot_library/Makefile [hunk 11]
```diff
@@ -20,6 +20,11 @@ TOP ?= $(shell pwd)/../../
 MAKEDIR ?= $(TOP)make/
 
 APP := robot_library
+VARIANTS = \
+	ellduino \
+	stm32f334nucleo \
+	stm32f4discovery
```
Comment (pekkanikander) [inline stm32/tests/robot_library/Makefile]: Please add the trailing backslash

File: stm32/tests/robot_library/robot_library.cpp [hunk 12]
```diff
@@ -17,6 +17,7 @@
  * along with ELL-i software.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <ellduino_spi.h>
```
Comment (pekkanikander) [inline stm32/tests/robot_library/robot_library.cpp]: This shouldn't be needed.  If it is, it is a bug somewhere else.  Please send me compiler output so that I can fix.

File: stm32/tests/robot_library/robot_library.cpp [hunk 13]
```diff
@@ -44,6 +45,7 @@ extern "C" {
 
     /*
      * C interface for the Arduino Digital API calls
+       Reorganise for Robot Framework integration
```
Comment (pekkanikander) [inline stm32/tests/robot_library/robot_library.cpp]: Please add the '*' at the right location to match the comment style.

File: stm32/variants/stm32f334nucleo/stm32f334nucleo_usart.c [hunk 14]
```diff
@@ -107,7 +107,7 @@ const SystemInitRecordData32Only USART_INIT_DefaultRecords[] = {
 };
 
 /*
- * Boot-time definitions for the STM32F334 series MCU USARTs.
+ * Boot-time definitions for the STM32F4 series MCU USARTs.
```
Comment (pekkanikander) [inline stm32/variants/stm32f334nucleo/stm32f334nucleo_usart.c:110]: This change is wrong.

File: stm32/cores/arduelli/Arduino_Serial.h (hunk 1 above)
Comment (asifsardar26) [inline stm32/cores/arduelli/Arduino_Serial.h]: The macros are no more needed after we have decided to put all the c-wrapper for c++ inside robot_library.cpp. These macros are also removed from system_init.h

Comment (pekkanikander) [inline stm32/cores/arduelli/Arduino_Serial.h]: They are not needed, but I want to preserve them.  Also, I did _NOT_ remove them from system_init.h.  That's one of the places where I edited your previous pull request.

Comment (asifsardar26) [inline stm32/cores/arduelli/Arduino_Serial.h]: Hi Pekka,

The changes are fixed.

May be these foolish changes arise because I was pulling the wrong remote
branch after my branch commits diverged from remote/origin branch. Now, my
branch is up-to-date with asifsardar26/feature-enc... branch. It is ahead
by 8 commits from origin/feacture-enc... branch because you haven't merged
my commits yet.

I am still grasping concepts related to remotes, branches and rebasing from
http://git-scm.com/book/en. I think I am very close to understand what
happens behind the git & github.

The pull request has few files left to review.

Moreover, I have noticed that the macros are not put in other places e.g.
in SPI header files. It would be good to replace them with macros.

On Fri, Sep 26, 2014 at 2:41 PM, Pekka Nikander notifications@github.com
wrote:

> In stm32/cores/arduelli/Arduino_Serial.h:
> 
> > @@ -59,7 +59,7 @@ class SerialClass : public Stream {
> >  constexpr SerialClass::SerialClass(const USART &usart)
> >      : usart_(usart) {}
> > 
> > -ARDUINO_INLINE_MEMBER_FUNCTION
> > +inline
> 
> They are not needed, but I want to preserve them. Also, I did _NOT_
> remove them from system_init.h. That's one of the places where I edited
> your previous pull request.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Ell-i/Runtime/pull/31/files#r18084765.

## 

_With Best Regards,Asif Sardar.+358 43 8265795_

File: stm32/emulator/inc/GPIO.h [hunk 15]
```diff
@@ -38,26 +46,31 @@ class GeneralPurposeInputOutput {
...
-    DEFINE_REGISTER(32, GPIO, AFR, 0)[2];
-#elif defined(__STM32F051__)
-    Register_GPIO_BSRR BSRR;
-    Register_GPIO_AFR AFR[2];
-    //Register_GPIO_LCKR LCKR;
-    Register_GPIO_BRR BRR;
-#else
-# error "Unknown MCU die.  Please define."
-#endif
+    #if defined(__STM32F407__) || defined(__STM32F334__)
```
Comment (pekkanikander) [inline stm32/emulator/inc/GPIO.h]: Again, please don't indent like this but like above

File: stm32/emulator/inc/GPIO.h [hunk 16]
```diff
@@ -38,26 +46,31 @@ class GeneralPurposeInputOutput {
...
+     Register_GPIO_BSRRL BSRRL;
```
Comment (pekkanikander) [inline stm32/emulator/inc/GPIO.h]: On this line and the following lines you have an extra space, intending too much

File: stm32/emulator/inc/GPIO.h [hunk 17]
```diff
@@ -38,26 +46,31 @@ class GeneralPurposeInputOutput {
...
+    # error "Unknown MCU die.  Please define."
+    #endif
 protected:
     GeneralPurposeInputOutput()
         : ODR(IDR)
-#if defined(__STM32F051__)
-        , BSRR(ODR)
-        , BRR(ODR)
-#endif
+        #if defined(__STM32F407__) || defined(__STM32F334__)
```
Comment (pekkanikander) [inline stm32/emulator/inc/GPIO.h]: Again, wrong type of indentation.

File: stm32/libraries/SPI/src/variants/stm32f334nucleo_spi.cpp [hunk 18]
```diff
@@ -47,4 +47,4 @@ struct SPIdynamicFields _SPI1dynamicFields;
  * XXX.  Linking hack, to be removed once linking gets fixed.
  */
 
-const void * __SPI1 = &SPI1_INIT;
+const void * __SPI1 = &SPI1_INIT;
```
Comment (pekkanikander) [inline stm32/libraries/SPI/src/variants/stm32f334nucleo_spi.cpp]: Add the missing newline.

**Comment by pekkanikander** [comment]
Asif, much better than before, thanks!

A few problems with indentation still.  Please fix those and it will be ready for merging.

**Comment by asifsardar26** [comment]
Hello Pekka,

Please check them now.

On Sun, Sep 28, 2014 at 5:31 PM, Pekka Nikander notifications@github.com
wrote:

> Asif, much better than before, thanks!
> 
> A few problems with indentation still. Please fix those and it will be
> ready for merging.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Ell-i/Runtime/pull/31#issuecomment-57087225.

## 

_With Best Regards,Asif Sardar.+358 43 8265795_

**Comment by pekkanikander** [comment]
Much better, thanks!

Two things still:

I don't remember the reason for adding the VARIANTS into the stm32/tests/robot_library/Makefile; please explain.  If it works without that addition, please revert the change.

And isn't the change to stm32/variants/stm32f334nucleo/stm32f334nucleo_usart.c wrong?  If so, please revert the change.

--Pekka

**Comment by asifsardar26** [comment]
The asifsardar26/Runtime/feature-enc.../stm32/variants/stm32f334nucleo/stm32f334nucleo_usart.c
matches with the Ell-i/Runtime/feature-enc.../stm32/variants/
stm32f334nucleo/stm32f334nucleo_usart.c.

Moreover, VARIANTS were not needed in the Makefile,

On Mon, Sep 29, 2014 at 5:27 PM, Pekka Nikander notifications@github.com
wrote:

> Much better, thanks!
> 
> Two things still:
> 
> I don't remember the reason for adding the VARIANTS into the
> stm32/tests/robot_library/Makefile; please explain. If it works without
> that addition, please revert the change.
> 
> And isn't the change to
> stm32/variants/stm32f334nucleo/stm32f334nucleo_usart.c wrong? If so, please
> revert the change.
> 
> --Pekka
> 
> On 2014–09–29, at 13:39 , Asif Sardar notifications@github.com wrote:
> 
> > Hello Pekka,
> > 
> > Please check them now.
> > 
> > On Sun, Sep 28, 2014 at 5:31 PM, Pekka Nikander <
> > notifications@github.com>
> > wrote:
> > 
> > > Asif, much better than before, thanks!
> > > 
> > > A few problems with indentation still. Please fix those and it will be
> > > ready for merging.
> > > 
> > > —
> > > Reply to this email directly or view it on GitHub
> > > https://github.com/Ell-i/Runtime/pull/31#issuecomment-57087225.
> > 
> > ## 
> > 
> > _With Best Regards,Asif Sardar.+358 43 8265795_
> > —
> > Reply to this email directly or view it on GitHub.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Ell-i/Runtime/pull/31#issuecomment-57168519.

## 

_With Best Regards,Asif Sardar.+358 43 8265795_
